File: src/eval/presets.ts
import {loadConfig, type PipelineConfig, type ChunkConfig, type SearchConfig} from "../pipeline/config.js"

export enum SweepLevel
{
    Fast,
    Medium,
    Thorough,
}

export function parseSweepLevel(s: string): SweepLevel
{
    if (s === "fast") return SweepLevel.Fast
    if (s === "medium") return SweepLevel.Medium
    if (s === "thorough") return SweepLevel.Thorough
    throw new Error(`unknown sweep level: ${s} (expected: fast | medium | thorough)`)
}

export function getPreset(level: SweepLevel): PipelineConfig[]
{
    if (level === SweepLevel.Fast) return FAST
    if (level === SweepLevel.Medium) return MEDIUM
    return THOROUGH
}

const base = loadConfig()

function withSearch(search: SearchConfig): PipelineConfig
{
    return { chunk: base.chunk, search }
}

function withChunk(chunk: ChunkConfig, search: SearchConfig): PipelineConfig
{
    return { chunk, search }
}

const FAST: PipelineConfig[] = [
    withSearch({ alpha: 0.7, k: 5 }),
]

const MEDIUM: PipelineConfig[] = [
    withSearch({ alpha: 0.0, k: 5 }),
    withSearch({ alpha: 0.3, k: 5 }),
    withSearch({ alpha: 0.5, k: 5 }),
    withSearch({ alpha: 0.7, k: 5 }),
    withSearch({ alpha: 1.0, k: 5 }),
    withSearch({ alpha: 0.5, k: 10 }),
    withSearch({ alpha: 0.7, k: 10 }),
    withSearch({ alpha: 0.5, k: 20 }),
    withSearch({ alpha: 0.7, k: 20 }),
]

const THOROUGH: PipelineConfig[] = [
    withChunk({ chunkSize: 400, overlap: 0 },  { alpha: 0.5, k: 5 }),
    withChunk({ chunkSize: 400, overlap: 0 },  { alpha: 0.7, k: 5 }),
    withChunk({ chunkSize: 400, overlap: 50 }, { alpha: 0.5, k: 5 }),
    withChunk({ chunkSize: 400, overlap: 50 }, { alpha: 0.7, k: 5 }),
    withChunk({ chunkSize: 600, overlap: 0 },  { alpha: 0.5, k: 5 }),
    withChunk({ chunkSize: 600, overlap: 0 },  { alpha: 0.7, k: 5 }),
    withChunk({ chunkSize: 600, overlap: 50 }, { alpha: 0.5, k: 5 }),
    withChunk({ chunkSize: 600, overlap: 50 }, { alpha: 0.7, k: 5 }),
    withChunk({ chunkSize: 800, overlap: 0 },  { alpha: 0.5, k: 5 }),
    withChunk({ chunkSize: 800, overlap: 0 },  { alpha: 0.7, k: 5 }),
    withChunk({ chunkSize: 800, overlap: 100 },{ alpha: 0.5, k: 5 }),
    withChunk({ chunkSize: 800, overlap: 100 },{ alpha: 0.7, k: 5 }),
]

File: arke/eval/presets.py
from dataclasses import replace
from itertools import product

from arke.server.config import Config

FAST = "fast"
MEDIUM = "medium"
THOROUGH = "thorough"


def get_preset(level: str, base: Config) -> list[Config]:
    if level == FAST:
        return _expand(base, chunk_sizes=[base.chunk_size], overlaps=[base.overlap], alphas=[base.alpha], ks=[base.k])
    if level == MEDIUM:
        return _expand(base, chunk_sizes=[base.chunk_size], overlaps=[base.overlap], alphas=[0.0, 0.3, 0.5, 0.7, 1.0], ks=[5, 10, 20])
    if level == THOROUGH:
        return _expand(base, chunk_sizes=[500, 1000], overlaps=[0.0, 0.2], alphas=[0.0, 0.3, 0.5, 0.7, 1.0], ks=[5, 10, 20])

    raise ValueError(f"unknown sweep level: {level} (expected: {FAST} | {MEDIUM} | {THOROUGH})")


def _expand(base: Config, chunk_sizes: list[int], overlaps: list[float], alphas: list[float], ks: list[int]) -> list[Config]:
    out: list[Config] = []
    for chunk_size, overlap, alpha, k in product(chunk_sizes, overlaps, alphas, ks):
        cfg = replace(base, chunk_size=chunk_size, overlap=overlap, alpha=alpha, k=k)
        out.append(cfg.resolved())
    return out
